### ISSUE TITLE
Android - Support Glide 4.x (e.g. 4.7.0)

### ISSUE THREAD

**Issue by hansemannn** [AndreaVitale/av.imageview#63]
We are using this library together with [Ti.ImagePicker](https://github.com/prashantsaini1/titanium-imagepicker) which uses Glide 4.7.0 (from April 2018). Unfortunately, this leads to incompatibility issues with this module, since this one still uses Glide 3.8.0. It looks like some classes have been moved in Glide 4.x (from May 2017), so migrating may take some effort.

**Comment by AndreaVitale** [comment]
Tomorrow I will start to work on this. I'll create a new branch and will merge when it will be finished.

**Comment by AndreaVitale** [comment]
Just for update: I'm working on it on branch https://github.com/AndreaVitale/av.imageview/tree/issue-63 integrating latest Glide version 4.9.0 atm. 
Is still in development but I solved also #62 and I changed how contentMode is handled so.. also #45 should be resolved.
I think that due to the new Glide structure we'll have problems to support the actual `validatesSecureCertificate` but.. step by step :-)

**Comment by phobeous** [comment]
I'll try to put an eye on it.

**Comment by hansemannn** [comment]
@AndreaVitale This is huge, THANK YOU!

**Comment by AndreaVitale** [comment]
Guys I finished a complete rewriting of the android module on [issue-63 branch](https://github.com/AndreaVitale/av.imageview/tree/issue-63).
It includes the latest stable Glide release (4.9.0) and it is really really really leaner then the previous one.

Please test the bundled module version (4.0.0) downloading it from [here](https://github.com/AndreaVitale/av.imageview/raw/issue-63/releases/av.imageview-android-4.0.0.zip) and let me know if you find any kind of issues for currently supported properties.

**Breaking changes**
- property `dontAnimate` is now named `animated` and is `true` by default
- property `rounded` must be used with `CONTENT_MODE_ASPECT_FIT` to be correctly rendered
- property `validatesSecureCertificate` (for allowing self-signed certificates) is not currently supported.

About the last point: sorry @phobeous. 
I was unable to create a custom `AppGlideModule`. If anyone can help on this is really appreciated.

**Comment by m1ga** [comment]
jsut tried a simple replace of the old module and two things I've found:
* setSignature() is missing (I use this for custom caching https://bumptech.github.io/glide/doc/caching.html#custom-cache-invalidation)
* got this error:
```
[WARN]  Glide: Load failed for /images/bg.jpg with size [1080x540]
[WARN]  Glide: class com.bumptech.glide.load.engine.GlideException: Failed to load resource
[WARN]  Glide: There was 1 cause:
[WARN]  Glide: java.net.MalformedURLException(no protocol: /images/bg.jpg)
[WARN]  Glide:  call GlideException#logRootCauses(String) for more detail
[WARN]  Glide:   Cause (1 of 1): class com.bumptech.glide.load.engine.GlideException: Fetching data failed, class java.io.InputStream, REMOTE
[WARN]  Glide: There was 1 cause:
[WARN]  Glide: java.net.MalformedURLException(no protocol: /images/bg.jpg)
[WARN]  Glide:  call GlideException#logRootCauses(String) for more detail
[WARN]  Glide:     Cause (1 of 1): class java.net.MalformedURLException: no protocol: /images/bg.jpg
```

same image is working in the current module. Didn't dig deeper there, just wanted to see if I can just switch modules

**Comment by hansemannn** [comment]
@AndreaVitale I just tried the module as well and many (remote) images do not load anymore. No error logs, only some `AvImageView: Destroying`.

**Comment by AndreaVitale** [comment]
About images that aren’t displayed... could you please provide some examples? I tried all images that are in example app.js and I didn’t find any issue.

**Comment by m1ga** [comment]
here is one image: http://migaweb.de/bg.jpg (placed in `app/assets/android/images/res-xxhdpi/`)

**Comment by hansemannn** [comment]
Any progress here? I'm happy to help!

**Comment by AndreaVitale** [comment]
@m1ga could you please try [this new](https://github.com/AndreaVitale/av.imageview/blob/issue-63/releases/av.imageview-android-4.0.0.zip) 4.0.0 version? I tried with density specific images and now is working as expected. Let me know!

**Comment by hansemannn** [comment]
Density specific images seem to work again. But unfortunately remote images with a `defaultImage` set still don't work.

**Comment by hansemannn** [comment]
I think #77 fixes this?

**Comment by AndreaVitale** [comment]
It is already merged in this branch, could you the latest archive in issue-63 branch?

**Comment by m1ga** [comment]
there is a typo:
```
[ERROR] Failed to compile Java source files:
[ERROR]   
[ERROR] warning: [options] bootstrap class path not set in conjunction with -source 1.7
[ERROR] av.imageview/android/src/av/imageview/AVImageView.java:30: error: class AvImageView is public, should be declared in a file named AvImageView.java
[ERROR] public class AvImageView extends TiUIView
[ERROR]        ^
[ERROR] Note: Some input files use or override a deprecated API.
[ERROR] Note: Recompile with -Xlint:deprecation for details.
[ERROR] Note: /home/miga/tools/av.imageview/android/src/av/imageview/AVImageView.java uses unchecked or unsafe operations.
[ERROR] Note: Recompile with -Xlint:unchecked for details.
[ERROR] 1 error
[ERROR] 1 warning
```
the v in the filename needs to be lowercase

and the same error:
```
[WARN]   Glide: Load failed for /images/avatar.jpg with size [289x289]
[WARN]   Glide: class com.bumptech.glide.load.engine.GlideException: Failed to load resource
[WARN]   Glide: There was 1 cause:
[WARN]   Glide: java.net.MalformedURLException(no protocol: /images/avatar.jpg)
[WARN]   Glide: call GlideException#logRootCauses(String) for more detail
[WARN]   Glide: Cause (1 of 1): class com.bumptech.glide.load.engine.GlideException: Fetching data failed, class java.io.InputStream, REMOTE
[WARN]   Glide: There was 1 cause:
[WARN]   Glide: java.net.MalformedURLException(no protocol: /images/avatar.jpg)
[WARN]   Glide: call GlideException#logRootCauses(String) for more detail
[WARN]   Glide: Cause (1 of 1): class java.net.MalformedURLException: no protocol: /images/avatar.jpg
[INFO]   Glide: Root cause (1 of 1)
[INFO]   Glide: java.net.MalformedURLException: no protocol: /images/avatar.jpg
[INFO]   Glide: at java.net.URL.<init>(URL.java:601)
[INFO]   Glide: at java.net.URL.<init>(URL.java:498)
[INFO]   Glide: at java.net.URL.<init>(URL.java:447)
[INFO]   Glide: at com.bumptech.glide.load.model.GlideUrl.getSafeUrl(GlideUrl.java:71)
[INFO]   Glide: at com.bumptech.glide.load.model.GlideUrl.toURL(GlideUrl.java:62)
[INFO]   Glide: at com.bumptech.glide.load.data.HttpUrlFetcher.loadData(HttpUrlFetcher.java:59)
[INFO]   Glide: at com.bumptech.glide.load.engine.SourceGenerator.startNext(SourceGenerator.java:62)
[INFO]   Glide: at com.bumptech.glide.load.engine.DecodeJob.runGenerators(DecodeJob.java:309)
[INFO]   Glide: at com.bumptech.glide.load.engine.DecodeJob.runWrapped(DecodeJob.java:279)
[INFO]   Glide: at com.bumptech.glide.load.engine.DecodeJob.run(DecodeJob.java:235)
[INFO]   Glide: at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
[INFO]   Glide: at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
[INFO]   Glide: at java.lang.Thread.run(Thread.java:764)
[INFO]   Glide: at com.bumptech.glide.load.engine.executor.GlideExecutor$DefaultThreadFactory$1.run(GlideExecutor.java:446)
```

**Comment by m1ga** [comment]
the problem only appears when using `.setImage()` or `.image = `. When I set the image in my tss file it doesn't throw that error.

**Comment by AndreaVitale** [comment]
@m1ga have you tried the latest dist in this branch? I forget to move the filename via git but the was build with the correct naming.

**Comment by m1ga** [comment]
yes, compiled the latest issue-63 branch. I've used rounded + ASPECT_FILL in some places which isn't working either now but that doesn't throw an error just isn't rounded (but I saw that it needs to be FIXED now)

**Comment by m1ga** [comment]
@AndreaVitale 
here you check for local and remote images:
https://github.com/AndreaVitale/av.imageview/blob/issue-63/android/src/av/imageview/AvImageView.java#L108
but here you don't:
https://github.com/AndreaVitale/av.imageview/blob/issue-63/android/src/av/imageview/ImageViewProxy.java#L94
so it tries to open my local image as a remote URL

**Comment by AndreaVitale** [comment]
@m1ga could you please provide the latest commit? Thanks!

**Comment by m1ga** [comment]
in one Listview I have those messages:
```
TiDrawableReference: (main) [764,83129] Could not open stream to get bitmap
TiDrawableReference: (main) [0,83129] Could not open stream to get bitmap
Glide: Load failed for null with size [-1x-1]
Glide: class com.bumptech.glide.load.engine.GlideException: Received null model
TiDrawableReference: (main) [668,83797] Could not open stream to get bitmap
TiDrawableReference: (main) [0,83797] Could not open stream to get bitmap
TiDrawableReference: (main) [668,83797] Could not open stream to get bitmap
TiDrawableReference: (main) [0,83797] Could not open stream to get bitmap
Glide: Load failed for null with size [0x0]
Glide: class com.bumptech.glide.load.engine.GlideException: Received null model
```
and some images won't load. I'll check that later

**Comment by m1ga** [comment]
` builder = builder.apply(options);` is missing here:
https://github.com/AndreaVitale/av.imageview/blob/issue-63/android/src/av/imageview/AvImageView.java#L187
(like in line 164)
that will show rounded images for local files again!

and I found the solution for the problem above:
https://github.com/AndreaVitale/av.imageview/blob/issue-63/android/src/av/imageview/AvImageView.java#L172
I have an Listview with images and I have to change that line above to use the `filename` parameter like this:
`Drawable imageDrawable = TiDrawableReference.fromUrl(this.proxy.get(), filename).getDrawable();`

now all images are visible in my app again!

**Comment by AndreaVitale** [comment]
Damn! Thanks @m1ga ! I applied your fixes, verified that in ListView I can apply round options for local images. Do you have other issues?

**Comment by m1ga** [comment]
I don't have any other issues but it looks like there are some other issues in #75 (even I don't have problems loading `/image/file.jpg`, `img.nativePath` and remote URLs. But will do some more tests tomorrow

**Comment by m1ga** [comment]
Looks like there is some bug in this setup:
* ScrollableView -> 10x Widget (containing an av.ImageView)

using the av.imageviews without a widget works fine or increasing the cacheSize of the scrollableview.

3.x is not having that issue. Will do some more debugging

**Comment by AndreaVitale** [comment]
Fixed in the released 4.0.0 version!
